### DISCUSSION TITLE
zip - FS error?

### DISCUSSION THREAD

**Discussion by nenge123** [gildas-lormeau/zip.js#359] [Q&A] [answered]
question
---
文件存在解压不完整.
例如我解压1MB的文本,存在只解压一半的问题.
超多文件时,字体文件残缺导致无法显示.

await zip.importBlob并不管用

**Comment by gildas-lormeau** [comment]
I added the missing `FS#importUint8Array` method in the last version.

Regarding the decompression issue, you have to provide me some code that I can run in order to debug it. Ideally, it should be as short as possible.

**Comment by nenge123** [comment]


**Reply by gildas-lormeau** [reply]
I'm not sure I followed correctly the instructions. The last step was not clear to me. However, when I click on the button "+" at the top of the page, I get the following error message in the console: `TypeError: can't access property "FS", zip.fs is undefined`. `zip` is defined though. Maybe you did not include the correct js file of zip.js?

**Comment by nenge123** [comment]


**Reply by gildas-lormeau** [reply]
I'm sorry but the procedure is a bit too complicated for me.

However, I tested your code almost as-is, I just added minimal implementations for `Nenge.unitl`, `ZipInitJS` and `T.Encoding`. The test seems to work as expected. Here are my test files.

```index.html```
```html
<script src="https://deno.land/x/zipjs/dist/zip-fs-full.js"></script>
<script src="test.js"></script>
<input type=file accept="application/zip" onchange="init(this)">
```

```test.js```
```js
async function init(input) {
  const arrayBuffer = await input.files[0].arrayBuffer();
  const uint8Array = new Uint8Array(arrayBuffer);
  await test(uint8Array);
}

async function test(uint8Array) {
  const Nenge = {
    unitl: {}
  };

  Object.entries({
    ZipInitJS: async () => { },
    T: {
      Encoding: "utf-8"
    },
    unZip: async function (u8, ARG = {}) {
      let {
        process,
        password,
        packtext
      } = ARG, F = this, T = F.T;
      await F.ZipInitJS();
      let zipFs = new zip.fs.FS(),
        contents = {};
      await zipFs['importBlob'](u8 instanceof Blob ? u8 : new Blob([u8]), {
        'filenameEncoding': T.Encoding
      });
      await Promise.all(
        zipFs.entries.map(async entry => {
          let data = entry.data;
          if (!data || data.directory) return;
          let opt = {
            'onprogress': (a, b) => process && process(packtext + entry.name + ':' + Math.ceil(a * 100 / b) + '%')
          };
          contents[data.filename] = await entry.getUint8Array(opt);
          return true;
        }));
      delete F.ZipPassword;
      return contents;
    }
  }).forEach(entry => {
    let opt;
    if (typeof entry[1] == 'string') opt = {
      get: () => entry[1]
    };
    else opt = {
      value: entry[1]
    };
    Object.defineProperty(Nenge.unitl, entry[0], opt);
  });
  
  const contents = await Nenge.unitl.unZip(uint8Array);
  console.log(Object.entries(contents).map(([key, value]) => [key, value.length]).join("\n"));
}
```

The test is also online here (please let me know which sites are not blocked by the GFW, if any):
 - https://jsfiddle.net/zxhn1ad8/36/
 - https://jsbin.com/cakefequjo/edit?html,console,output
 - https://plnkr.co/edit/CeqyyCr1ItGQLsjj?open=lib%2Fscript.js&preview


I did some tests with multiple zip files. The `contents` object always contains complete `Uint8Array` objects (see the log in the console).

**Reply by gildas-lormeau** [reply]
Are you able to open any of these URLs without being blocked?
- https://jsfiddle.net/zxhn1ad8/36/
- https://jsbin.com/cakefequjo/edit?html,console,output
- https://plnkr.co/edit/CeqyyCr1ItGQLsjj?open=lib%2Fscript.js&preview

If yes, which ones?

**Reply by gildas-lormeau** [reply]
That will be very complicated for me to help you if all of these tools are blocked... Don't you have some alternative tool (a site where we can run small tests like this one) accessible in China that we could use together?

I did another test (see below, same HTML file) using a `Blob` and displaying the number of "empty" characters (character 0 in the ASCII table). The results looks fine to me. For example, I always see 0 as result for entries  containing text files.

```js
async function init(input) {
  const blob = input.files[0];
  await test(blob);
}

async function test(blob) {
  const Nenge = {
    unitl: {}
  };
  Object.entries({
    ZipInitJS: async () => {},
    T: {
      Encoding: "utf-8"
    },
    unZip: async function(u8, ARG = {}) {
      let {
        process,
        password,
        packtext
      } = ARG, F = this, T = F.T;
      await F.ZipInitJS();
      let zipFs = new zip.fs.FS(),
        contents = {};
      await zipFs['importBlob'](u8 instanceof Blob ? u8 : new Blob([u8]), {
        'filenameEncoding': T.Encoding
      });
      await Promise.all(
        zipFs.entries.map(async entry => {
          let data = entry.data;
          if (!data || data.directory) return;
          let opt = {
            'onprogress': (a, b) => process && process(packtext + entry.name + ':' + Math.ceil(a * 100 / b) + '%')
          };
          contents[data.filename] = await entry.getUint8Array(opt);
          return true;
        }));
      delete F.ZipPassword;
      return contents;
    }
  }).forEach(entry => {
    let opt;
    if (typeof entry[1] == 'string') opt = {
      get: () => entry[1]
    };
    else opt = {
      value: entry[1]
    };
    Object.defineProperty(Nenge.unitl, entry[0], opt);
  })
  const contents = await Nenge.unitl.unZip(blob);
  console.log(
    Object.entries(contents).map(([key,  value]) => [key, value.filter(value => value === 0).length]).join("\n")
  );
}
```

For the record, it can be run from here: https://jsbin.com/faxajitoju/1/edit?html,console,output

**Reply by gildas-lormeau** [reply]
I think I can reproduce your issue in Chrome actually. It would be related to Compression Streams. Calling `zip.configure({ useCompressionStream: false });` seems to circumvent this issue. It seems also it would be specific to the `FS` API which is more unexpected... I will take a look at this tomorrow because it's late here.

**Comment by nenge123** [comment]
@gildas-lormeau 

OK you can test this link,upload you local zip-fs.js to run!


[https://retroarch.nenge.net/zipfs.html](https://retroarch.nenge.net/zipfs.html)

**Comment by gildas-lormeau** [comment, marked as answer]
The bug leading to incomplete files should be fixed in the version 2.6.20 that I just published.
